gnpy/core/science_utils.py

```diff
@@ -0,0 +1,842 @@
+
+        return np.vstack(dpdz)
+
+class NliSolver:
+    """ This class implements the NLI models.
+        Model and method can be specified in `self.nli_params.method`.
+        List of implemented methods:
+        'gn_model_analytic': brute force triple integral solution
+        'ggn_spectrally_separated_xpm_spm': XPM plus SPM
+    """
+
+    def __init__(self, nli_params=None, fiber_params=None):
+        """ Initialize the fiber object with its physical parameters
+        """
+        self.fiber_params = fiber_params
+        self.nli_params = nli_params
+        self.stimulated_raman_scattering = None
+
+    @property
+    def fiber_params(self):
+        return self._fiber_params
+
+    @fiber_params.setter
+    def fiber_params(self, fiber_params):
+        self._fiber_params = fiber_params
+
+    @property
+    def stimulated_raman_scattering(self):
+        return self._stimulated_raman_scattering
+
+    @stimulated_raman_scattering.setter
+    def stimulated_raman_scattering(self, stimulated_raman_scattering):
+        self._stimulated_raman_scattering = stimulated_raman_scattering
+
+    @property
+    def nli_params(self):
+        return self._nli_params
+
+    @nli_params.setter
+    def nli_params(self, nli_params):
+        """
+        :param model_params: namedtuple containing the parameters used to compute the NLI.
+        """
+        self._nli_params = nli_params
+
+    def compute_nli(self, carrier, *carriers):
+        """ Compute NLI power generated by the WDM comb `*carriers` on the channel under test `carrier`
+        at the end of the fiber span.
+        """
+        if 'gn_model_analytic' == self.nli_params.nli_method_name.lower():
+            carrier_nli = self._gn_analytic(carrier, *carriers)
+        elif 'ggn_spectrally_separated' in self.nli_params.nli_method_name.lower():
+            eta_matrix = self._compute_eta_matrix(carrier, *carriers)
+            carrier_nli = self._carrier_nli_from_eta_matrix(eta_matrix, carrier, *carriers)
+        else:
+            raise ValueError(f'Method {self.nli_params.method_nli} not implemented.')
+
+        return carrier_nli
+
+    @staticmethod
+    def _carrier_nli_from_eta_matrix(eta_matrix, carrier, *carriers):
+        carrier_nli = 0
+        for pump_carrier_1 in carriers:
+            for pump_carrier_2 in carriers:
+                carrier_nli += eta_matrix[pump_carrier_1.channel_number-1, pump_carrier_2.channel_number-1] * \
+                               pump_carrier_1.power.signal * pump_carrier_2.power.signal
+        carrier_nli *= carrier.power.signal
+
+        return carrier_nli
+
+    def _compute_eta_matrix(self, carrier_cut, *carriers):
+        cut_index = carrier_cut.channel_number - 1
+        # Matrix initialization
+        matrix_size = max(carriers, key=lambda x: getattr(x, 'channel_number')).channel_number
+        eta_matrix = np.zeros(shape=(matrix_size, matrix_size))
+
+        # SPM
+        if self.nli_params.verbose:
```

Comment:
This (and everywhere else) should be converted to use the `logging` module. That way, the code just emits logging messages with a given severity (error/warn/info/debug/trace) and "component" (JSON parsing, Raman solver, PCE, etc), and the framework selects what to capture and what messages to discard.